add typehints with mypy

rwadman opened the pull request:
Hello! 
We're using flatten-dict in a typed context at work, so I created some stubs for it the other day. Felt like it could be nice to share them with you, but I think it's much easier to maintain typehints if they're written inline with the code, so I spent some time this weekend writing that up.

Had to upgrade black to avoid some broken import, and did my best to configure static type checking with mypy.

Don't know if you're interested at all, so feel free to ignore, but here you have it :)

rwadman committed:
add typehints with mypy

rwadman commented:
Noticed now that I forgot to add the type checking to CI, will see if I can figure that out...

rwadman committed:
add typechecking to CI

ianlini commented:
Thanks. I'm very interested in mypy, but didn't have time to learn it. I will spend some time understanding this PR.

ianlini commented:
Are you using the comments because we are still supporting Python 2.7? I think it's time to remove it so that we can use the type annotation.

rwadman commented:
Nice! Yes, the type comments are there to support python < 3.5, which (I think) is the first version with support for proper annotations.

Which python versions we should support is your call I feel. If your certain about dropping 2.7 (and probably also anything else before 3.5) I could update the PR to move away from the type comment style. 
It would also simplify the tox setup a bit since we would then be able to typecheck for each version of python with the tox environment for that version (to typecheck for python 2.7 you have to run mypy in python >3.5).

rwadman committed:
better ci script

rwadman committed:
run tests for py2.7 as include rather than matrix

ianlini commented:
Yes, please drop Python 2.7.

rwadman committed:
remove py27 from tox and CI

rwadman committed:
drop support for python2.7 and python <3.5

rwadman committed:
better alignment between type hints and implementation and typehints in tests

A bit concerned that the callables and generics are becoming too unwieldy though.

rwadman committed:
loosen up typevars

rwadman committed:
remove spurious comment
